modules/nf-core/gsea/gsea/main.nf

```diff
@@ -0,0 +1,77 @@
+    tuple val(meta), path("*/ranked_gene_list*.tsv")                                                                                        , emit: ranked_gene_list
+    tuple val(meta), path("*/gene_set_sizes.tsv")                                                                                           , emit: gene_set_sizes
+    tuple val(meta), path("*/butterfly_plot.png")                                                                                           , emit: butterfly_plot
+    tuple val(meta), path("*/global_es_histogram.png")                                                                                      , emit: histogram
+    tuple val(meta), path("*/heat_map_1.png")                                                                                               , emit: heatmap
+    tuple val(meta), path("*/pvalues_vs_nes_plot.png")                                                                                      , emit: pvalues_vs_nes_plot
+    tuple val(meta), path("*/ranked_list_corr_2.png")                                                                                       , emit: ranked_list_corr
+    tuple val(meta), path("*/[!gene_set_size|gsea_report|ranked_gene_list]*.tsv")                                                           , emit: gene_set_tsv, optional: true
+    tuple val(meta), path("*/[!gsea_report|heat_map_corr_plot|index|pos_snapshot|neg_snapshot]*.html")                                      , emit: gene_set_html, optional: true
+    tuple val(meta), path("*/[!butterfly|enplot|global_es_histogram|gset_rnd_es_dist|heat_map|pvalues_vs_nes_plot|ranked_list_corr]*.png")  , emit: gene_set_heatmap, optional: true
+    tuple val(meta), path("*/*_snapshot*.html")                                                                                             , emit: snapshot, optional: true
+    tuple val(meta), path("*/enplot*.png")                                                                                                  , emit: gene_set_enplot, optional: true
+    tuple val(meta), path("*/gset_rnd_es_dist*.png")                                                                                        , emit: gene_set_dist, optional: true
+    tuple val(meta), path("*/*.zip")                                                                                                        , emit: archive, optional: true
+    path "versions.yml"                                                                                                                     , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def args2 = task.ext.args2 ?: [ ]
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    # Run GSEA
+    gsea-cli GSEA \
+        -res $gct \
+        -cls ${cls}#${args2.target}_versus_${args2.reference} \
```

Comment:
I still don't like this, but at least should be reversed so this is just `args` and then `args2` is for the 'raw' user stuff.

Is there a reason why you can't just hardcode `target` and `reference` in the name, but carry that information in the `meta map` (and then e.g. publishDir those files in a directory named after those, or even use the meta info  to insert into `prefix`?

I know you  really want to be able to refer to it, but I'm worried again it is not portable.

Or maybe even just `args`, `args2`, `args3`? I'm not really sure exactly where to document this though...

oooooooooooorrrrrrrrrrrrrrrrrrrrrr

we allow a `val` input channel exception here ;)

Reply:
We do that in sarek for some specific args/files combo:
https://github.com/nf-core/sarek/blob/master/conf/igenomes.config#L24

Reply:
So you would think it's acceptable?

Reply:
Ok to summarise slack @pinin4fjords 

- Will the tool fail execute if you don't supply the `#reference_versus_target` string : if so use `input` channels to define `reference` and `target`
- If the tool can run without them: you should leave it to the pipeline developer to pass whatever string they want via an `args`

Reply:
> * Will the tool fail execute if you don't supply the `#reference_versus_target` string : if so use `input` channels to define `reference` and `target`
> * If the tool can run without them: you should leave it to the pipeline developer to pass whatever string they want via an `args`

To reiterate from the [slack thread](https://nfcore.slack.com/archives/CJRH30T6V/p1670481521205589), yes, it will fail unless the .cls file has only two classes, which would be hard to ensure. So we're going with a separate channel to provide the reference and target.